homeassistant/components/advantage_air/climate.py

```diff
@@ -82,39 +83,116 @@ async def async_setup_entry(
 
+    _attr_fan_modes = [FAN_LOW, FAN_MEDIUM, FAN_HIGH]
     _attr_temperature_unit = TEMP_CELSIUS
     _attr_target_temperature_step = PRECISION_WHOLE
     _attr_max_temp = 32
     _attr_min_temp = 16
-    _attr_fan_modes = [FAN_AUTO, FAN_LOW, FAN_MEDIUM, FAN_HIGH]
-    _attr_hvac_modes = AC_HVAC_MODES
-    _attr_supported_features = (
-        ClimateEntityFeature.TARGET_TEMPERATURE | ClimateEntityFeature.FAN_MODE
-    )
 
     def __init__(self, instance, ac_key):
         """Initialize an AdvantageAir AC unit."""
         super().__init__(instance, ac_key)
-        if self._ac.get("myAutoModeEnabled"):
-            self._attr_hvac_modes = AC_HVAC_MODES + [HVACMode.AUTO]
+        self._attr_unique_id = f'{self.coordinator.data["system"]["rid"]}-{ac_key}'
+
+        self._attr_preset_modes = [ADVANTAGE_AIR_MYZONE]
+        self._attr_supported_features = ClimateEntityFeature.FAN_MODE
+
+        # Add "MyTemp" preset if available
+        if ADVANTAGE_AIR_MYTEMP_ENABLED in self._ac:
+            self._attr_preset_modes += [ADVANTAGE_AIR_MYTEMP]
+            self._attr_supported_features |= ClimateEntityFeature.PRESET_MODE
+
+        # Add "MyAuto" preset if available
+        if ADVANTAGE_AIR_MYAUTO_ENABLED in self._ac:
+            self._attr_preset_modes += [ADVANTAGE_AIR_MYAUTO]
+            self._attr_supported_features |= ClimateEntityFeature.PRESET_MODE
 
     @property
-    def target_temperature(self):
+    def target_temperature(self) -> float | None:
         """Return the current target temperature."""
         return self._ac["setTemp"]
 
     @property
-    def hvac_mode(self):
+    def hvac_mode(self) -> HVACMode:
         """Return the current HVAC modes."""
         if self._ac["state"] == ADVANTAGE_AIR_STATE_ON:
-            return ADVANTAGE_AIR_HVAC_MODES.get(self._ac["mode"])
+            return ADVANTAGE_AIR_HVAC_MODES[self._ac["mode"]]
         return HVACMode.OFF
 
     @property
-    def fan_mode(self):
+    def hvac_modes(self) -> list[HVACMode]:
+        """Return the available HVAC modes."""
+        # MyTemp only supports cooling and heating
+        if self._ac.get(ADVANTAGE_AIR_MYTEMP_ENABLED):
+            return [HVACMode.OFF, HVACMode.COOL, HVACMode.HEAT]
+        # MyAuto adds support for Auto
+        if self._ac.get(ADVANTAGE_AIR_MYAUTO_ENABLED):
+            return [
+                HVACMode.OFF,
+                HVACMode.COOL,
+                HVACMode.HEAT,
+                HVACMode.FAN_ONLY,
+                HVACMode.DRY,
+                HVACMode.HEAT_COOL,
+            ]
+        # MyZone does not support auto
+        return [
+            HVACMode.OFF,
+            HVACMode.COOL,
+            HVACMode.HEAT,
+            HVACMode.FAN_ONLY,
+            HVACMode.DRY,
+        ]
+
+    @property
+    def fan_mode(self) -> str | None:
         """Return the current fan modes."""
         return ADVANTAGE_AIR_FAN_MODES.get(self._ac["fan"])
 
+    @property
+    def fan_modes(self) -> list[str] | None:
```

Comment:
Can you set this as `self._attr_fan_modes`?

Reply:
No because the available fan modes can change depending on the setting on the device.

Reply:
@bdraco is this also unacceptable?

Reply:
Correct, fan modes aren't allowed to change at runtime.

Reply:
@bdraco is this appropriate?
https://github.com/Bre77/core/blob/aa-autoselect2/homeassistant/components/advantage_air/climate.py#L189-L192

Reply:
Yes. The message gets propagated so its a good idea to give the user some feedback.

Some examples
```
% git grep 'raise ValueError' */climate.py
airtouch4/climate.py:            raise ValueError(f"Unsupported HVAC mode: {hvac_mode}")
airtouch4/climate.py:            raise ValueError(f"Unsupported fan mode: {fan_mode}")
airtouch4/climate.py:            raise ValueError(f"Unsupported HVAC mode: {hvac_mode}")
airtouch4/climate.py:            raise ValueError(f"Unsupported fan mode: {fan_mode}")
balboa/climate.py:            raise ValueError(f"{preset_mode} is not a valid preset mode")
balboa/climate.py:            raise ValueError(f"{mode} can only be reported but not set")
deconz/climate.py:            raise ValueError(f"Unsupported fan mode {fan_mode}")
deconz/climate.py:            raise ValueError(f"Unsupported HVAC mode {hvac_mode}")
deconz/climate.py:            raise ValueError(f"Unsupported preset mode {preset_mode}")
deconz/climate.py:            raise ValueError(f"Expected attribute {ATTR_TEMPERATURE}")
ecobee/climate.py:            raise ValueError(
econet/climate.py:            raise ValueError(f"{hvac_mode} is not a valid mode.")
freedompro/climate.py:            raise ValueError(f"Got unsupported hvac_mode {hvac_mode}")
generic_thermostat/climate.py:                raise ValueError(f"Sensor has illegal state {state.state}")
generic_thermostat/climate.py:            raise ValueError(
gree/climate.py:            raise ValueError(f"Missing parameter {ATTR_TEMPERATURE}")
gree/climate.py:            raise ValueError(f"Invalid hvac_mode: {hvac_mode}")
gree/climate.py:            raise ValueError(f"Invalid preset mode: {preset_mode}")
gree/climate.py:            raise ValueError(f"Invalid fan mode: {fan_mode}")
gree/climate.py:            raise ValueError(f"Invalid swing mode: {swing_mode}")
maxcube/climate.py:            raise ValueError(f"unsupported HVAC mode {hvac_mode}")
maxcube/climate.py:            raise ValueError(
maxcube/climate.py:            raise ValueError(f"unsupported preset mode {preset_mode}")
melcloud/climate.py:            raise ValueError(f"Invalid hvac_mode [{hvac_mode}]")
melcloud/climate.py:            raise ValueError(
melcloud/climate.py:            raise ValueError(
melcloud/climate.py:            raise ValueError(f"Invalid hvac_mode [{hvac_mode}]")
mqtt/climate.py:        raise ValueError("preset_modes must not include preset mode 'none'")
nobo_hub/climate.py:            raise ValueError(
plugwise/climate.py:            raise ValueError("Invalid temperature requested")
radiotherm/climate.py:            raise ValueError(f"{fan_mode} is not a valid fan mode")
radiotherm/climate.py:            raise ValueError(f"{preset_mode} is not a valid preset_mode")
screenlogic/climate.py:            raise ValueError(f"Expected attribute {ATTR_TEMPERATURE}")
sensibo/climate.py:            raise ValueError("No target temperature provided")
vicare/climate.py:            raise ValueError("Cannot set hvac mode when vicare_modes are not known")
vicare/climate.py:            raise ValueError(f"Cannot set invalid hvac mode: {hvac_mode}")
vicare/climate.py:            raise ValueError(
vicare/climate.py:            raise ValueError(f"Cannot set invalid vicare mode: {vicare_mode}.")
whirlpool/climate.py:            raise ValueError(f"Invalid hvac mode {hvac_mode}")
whirlpool/climate.py:            raise ValueError(f"Invalid fan mode {fan_mode}")
yolink/climate.py:            raise ValueError(f"Received an invalid hvac mode: {hvac_mode}")
zwave_js/climate.py:            raise ValueError("Value requested is not available")
zwave_js/climate.py:            raise ValueError(f"Received an invalid fan mode: {fan_mode}") from None
zwave_js/climate.py:            raise ValueError(f"Received an invalid hvac mode: {hvac_mode}")
zwave_js/climate.py:            raise ValueError(f"Received an invalid preset mode: {preset_mode}")
```